skfda/representation/basis/_grid_basis.py

```diff
@@ -0,0 +1,60 @@
+from __future__ import annotations
+
+from typing import Any, TypeVar
+
+from ..._utils import _to_grid_points
+from ...typing._base import GridPointsLike
+from ...typing._numpy import NDArrayFloat
+from ._basis import Basis
+
+T = TypeVar("T", bound="_GridBasis")
+
+
+class _GridBasis(Basis):
+    """
+    Basis associated to a grid of points.
+
+    Used to express functions whose values are known in a grid of points.
+    Each basis function is guaranteed to be zero in all points of the grid
+    except one, where it is one. The intermediate values are interpolated
+    depending on the interpolation object passed as argument.
+
+    This basis is used internally as an alternate representation of a
+    FDataGrid object. In certain cases, it is more convenient to work with
+    FDataGrid objects as a basis representation in this basis since, then,
+    all functional data can be represented as an FDataBasis object.
+
+    Parameters:
+        grid_points: Grid of points where the functions are evaluated.
+
+    """
+
+    def __init__(
+        self,
+        *,
+        grid_points: GridPointsLike,
+    ) -> None:
+        """Basis constructor."""
+        self.grid_points = _to_grid_points(grid_points)
+        domain_range = tuple((s[0], s[-1]) for s in self.grid_points)
+        super().__init__(
+            domain_range=domain_range,
+            n_basis=len(self.grid_points[0]),
+        )
+
+    def _evaluate(self, eval_points: NDArrayFloat) -> NDArrayFloat:
+        raise NotImplementedError(
+            "Evaluation is not implemented in this basis",
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        return (
+            super().__eq__(other)
+            and self.grid_points == other.grid_points
+        )
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}(grid_points={self.grid_points}) "
+
+    def __hash__(self) -> int:
+        return hash((super(), f"{self.grid_points}"))
```

Comment:
Is `grid_points` hashable?

Reply:
@vnmabus,
By itself no. But converted to a string it is no longer mutable so I would hope so. I might be missing something but that's why it's in an f-string.

Reply:
Uh-uh, that is a bit fishy. I would convert it to tuples instead or something like that (see what `BSplineBasis` does with the `knots`).

In the long term, I think that the `hash` for the bases should not depend on the attributes, as the class is mutable.

Reply:
I would consider that changing the `grid_points` results in a different grid basis. That is why I included it in the hash. I think it defines the identity of the basis.  
I have changed it to use a tuple instead of convertin to string

Reply:
That is true. However, we only need that objects that compare equal have the same hash. The opposite, although desirable for performance reasons, is not required. Note that our only reason for having hashable bases in the first case is because they are an attribute for the `FDataBasis` ExtensionDType, and pandas requires that those are hashable.

I think that in the long term we should just hash the class type itself or something like that.

Reply:
For dictionaries, only the keys need to be hashable. Normally, mutable objects such as lists or arrays are not made hashable on purpose.

Reply:
Yeah, you are right. I was also thinking of sets, and I got them mixed up. 
To me, `GridBasis` does seem like it should be an immutable object. And it is fully characterized by the discretization points. With that said, since we do not have a need to implement effective hash methods, I think it is not unreasonable just to use the basis type as a hash. But then the method would be basically returning a constant. 

Personally, I'd rather change the `grid_points` to a property, making the object immutable, and keep the points in the hash. But I think the other option also makes sense. In any case, let me know what you think because I want to get to closing this PR rather sooner than later. The PLS functionality depends on this, and I'd like to have a clean PLS  PR at this point (with these commits integrated into develop).

Reply:
I think I will accept this PR in its current state. If we change the hash behaviour for basis objects, that should be done in another PR.

For me, at the beginning I though of basis objects as immutable too. However, they were several disadvantages with that approach, and I am reconsidering it now. First, Python does not have first-class support for custom immutable objects, apart from frozen dataclasses. Second, and more important, the current approach makes difficult or even impossible the choice of basis parameters in a `BasisSmoother` transformer using hyperparameter selection tools such as `GridSearchCV`. For example, for choosing the number of elements in the basis, one has to create different `Basis` objects with the appropriate parameter set. Choosing several basis parameters at the same time is too complex for average users.

Forgetting the idea of immutable basis, we can make `Basis` inherit from `BaseEstimator`, thus allowing it to play well with the hyperparameter selection tools in scikit-learn. I think we should investigate this approach further.